Packages/com.unity.inputsystem/InputSystem/Devices/ScreenKeyboard.cs

```diff
@@ -0,0 +1,306 @@
+        /// </summary>
+        public string initialText { get => m_InitialText; set => m_InitialText = value; }
+
+        /// <summary>
+        /// Placeholder text when input field text is empty.
+        /// </summary>
+        public string placeholderText { get => m_PlaceholderText; set => m_PlaceholderText = value; }
+
+        /// <summary>
+        /// Is auto correction applied?
+        /// </summary>
+        public bool autocorrection { get => m_Autocorrection; set => m_Autocorrection = value; }
+
+        /// <summary>
+        /// Can more than one line of text be entered in the input field ?
+        /// </summary>
+        public bool multiline { get => m_Multiline; set => m_Multiline = value; }
+
+        /// <summary>
+        /// Is the text masked (for passwords, etc)?
+        /// </summary>
+        public bool secure { get => m_Secure; set => m_Secure = value; }
+
+        /// <summary>
+        /// Is the keyboard opened in alert mode?
+        /// Note: Only available on iOS
+        /// </summary>
+        public bool alert { get => m_Alert; set => m_Alert = value; }
+
+        /// <summary>
+        /// Is input field hidden?
+        /// </summary>
+        public bool inputFieldHidden { get => m_InputFieldHidden; set => m_InputFieldHidden = value; }
+    }
+
+    /// <summary>
+    /// ScreenKeyboard base class for platform specific implementation.
+    ///
+    /// Known issues:
+    /// * There's a screen keyboard behavior differences in regards to input and focus loss:
+    ///   - [Android] When screen keyboard is shown, the main Unity window stops receiving input events, when clicking on Unity window, screen keyboard looses focus and closes itself
+    ///     (While it's possible to keep screen keyboard opened on focus loss, currently there's no way to make Unity window keep receiving input events like touch while screen keyboard is shown,
+    ///      this matches the behavior in old TouchScreenKeyboard implementation)
+    ///   - [iOS] When screen keyboard is shown, the main Unity window will continue receiving input events, when clicking on Unity window, screen keyboard will continue to be shown
+    ///   - [WSA] When screen keyboard is shown, the main Unity window will continue receiving input events, when clicking on Unity window, screen keyboard will continue to be shown
+    /// </summary>
+    internal abstract class ScreenKeyboard
+    {
+        protected ScreenKeyboardState m_KeyboardState;
+        protected ScreenKeyboardShowParams m_ShowParams;
+
+        private InlinedArray<Action<ScreenKeyboardState>> m_StatusChangedListeners;
+        private InlinedArray<Action<string>> m_InputFieldTextListeners;
+        private InlinedArray<Action<RangeInt>> m_SelectionChangedListeners;
+
+        /// <summary>
+        /// Subscribe to an event which is fired whenever screen keyboard state changes
+        /// </summary>
+        public event Action<ScreenKeyboardState> stateChanged
```

Comment:
Hmph, I think it makes sense what you're saying. I'll check

Reply:
As discussed, you can now pass callbacks to Show method, and to make it less chance for user to make a mistake, I turned callbacks to be delegates instead of events